compiler/qsc_codegen/src/qsharp.rs

```diff
@@ -0,0 +1,759 @@
+    let mut gen = QSharpGen::new(output);
+    for package in packages {
+        gen.visit_package(package);
+    }
+}
+
+pub fn write_store<W: Write>(output: W, store: &PackageStore) {
+    let mut gen = QSharpGen::new(output);
+    for (_, unit) in store {
+        gen.visit_package(&unit.ast.package);
+    }
+}
+
+#[must_use]
+pub fn write_store_string(store: &PackageStore) -> Vec<String> {
+    let mut package_strings: Vec<_> = vec![];
+    for (_, unit) in store {
+        package_strings.push(write_package_string(&unit.ast.package));
+    }
+    package_strings
+}
+
+#[must_use]
+pub fn write_package_string(package: &Package) -> String {
+    let mut output = Vec::new();
+    write(&mut output, &[package]);
+    let s = match std::str::from_utf8(&output) {
+        Ok(v) => v.to_owned(),
+        Err(e) => format!("Invalid UTF-8 sequence: {e}"),
+    };
+
+    output.clear();
+    format_str(&s)
+}
+
+struct QSharpGen<W: Write> {
+    pub(crate) output: W,
+}
+
+impl<W> QSharpGen<W>
+where
+    W: Write,
+{
+    pub fn new(output: W) -> Self {
+        Self { output }
+    }
+
+    pub fn write(&mut self, args: &str) {
+        let _ = write!(&mut self.output, "{args}");
+    }
+
+    pub fn writeln(&mut self, args: &str) {
+        self.write(args);
+        self.write("\n");
+    }
+
+    /// special case for tuple with one element
+    /// otherwise we are changing the semantics of the program
+    fn ensure_trailing_comma_for_arity_one_tuples<T>(&mut self, most: &[T]) {
+        if most.is_empty() {
+            self.write(",");
+        }
+    }
+}
+
+impl<W: Write> Visitor<'_> for QSharpGen<W> {
+    fn visit_package(&mut self, package: &'_ Package) {
+        package.nodes.iter().for_each(|n| match n {
+            TopLevelNode::Namespace(ns) => {
+                self.visit_namespace(ns);
+            }
+            TopLevelNode::Stmt(stmt) => self.visit_stmt(stmt),
+        });
+        package.entry.iter().for_each(|e| self.visit_expr(e));
+    }
+
+    fn visit_namespace(&mut self, namespace: &'_ Namespace) {
+        self.write("namespace ");
+        self.visit_ident(&namespace.name);
+        self.writeln("{");
+        namespace.items.iter().for_each(|i| {
+            self.visit_item(i);
+        });
+        self.write("}");
+    }
+
+    fn visit_item(&mut self, item: &'_ Item) {
+        item.attrs.iter().for_each(|a| self.visit_attr(a));
+        item.visibility
+            .iter()
+            .for_each(|v| self.visit_visibility(v));
+        match &*item.kind {
+            ItemKind::Err => {
+                unreachable!()
+            }
+            ItemKind::Callable(decl) => self.visit_callable_decl(decl),
+            ItemKind::Open(ns, alias) => {
+                self.write("open ");
+                self.visit_ident(ns);
+                for a in alias {
+                    self.write(" as ");
+                    self.visit_ident(a);
+                }
```

Comment:
This for loop confused me into thinking that there could be multiple aliases. Could you use a more Option-specific syntax here like:
```suggestion
                if let Some(a) = alias {
                    self.write(" as ");
                    self.visit_ident(a);
                }
```

Reply:
I guess I'm seeing this for loop pattern for options pretty consistently in this code. Is there a particular reason for doing that over the if-let-some pattern?